etna/datasets/internal_datasets.py

```diff
@@ -239,6 +257,88 @@ def get_m4_dataset(dataset_dir: Path, dataset_freq: str) -> None:
+    Competition. The M3 dataset consists of time series of yearly, quarterly, monthly and other data.
+    Each frequency mode has its own specific prediction horizon: 6 for yearly, 8 for quarterly,
+    18 for monthly, and 8 for other.
+
+    Parameters
+    ----------
+    dataset_dir:
+        The path for saving dataset locally.
+    dataset_freq:
+        Frequency mode.
+
+    References
+    ----------
+    .. [1] https://forvis.github.io/datasets/m3-data/
+    .. [2] https://forecasters.org/resources/time-series-data/m3-competition/
+    """
+    get_freq = {"monthly": "M", "quarterly": "Q-DEC", "yearly": "A-DEC", "other": "Q-DEC"}
+    get_horizon = {"monthly": 18, "quarterly": 8, "yearly": 6, "other": 8}
+    url_data = "https://forvis.github.io/data"
+    end_date = "2022-01-01"
+    freq = get_freq[dataset_freq]
+
+    dataset_dir.mkdir(exist_ok=True, parents=True)
+
+    data = pd.read_csv(f"{url_data}/M3_{dataset_freq}_TSTS.csv")
+
+    df_full = pd.DataFrame()
+    df_train = pd.DataFrame()
+    df_test = pd.DataFrame()
+
+    df_full_exog = pd.DataFrame()
+    df_test_exog = pd.DataFrame()
+    horizon = get_horizon[dataset_freq]
+    for _, group in data.groupby("series_id"):
+        group_copy = group.copy()
+        timestamps = pd.date_range(end=end_date, freq=freq, periods=group.shape[0])
+        group_copy.rename(columns={"timestamp": "origin_timestamp"}, inplace=True)
+        group_copy.rename(columns={"series_id": "segment"}, inplace=True)
+        group_copy.rename(columns={"value": "target"}, inplace=True)
+        group_copy["segment"] = group_copy.apply(lambda x: x["segment"] + "_" + x["category"], axis=1)
+        group_copy.drop(columns=["category"], inplace=True)
+        group_copy["timestamp"] = timestamps
+
+        df_full_part_exog = group_copy.copy()
+        df_full_part_exog.drop(columns=["target"], inplace=True)
+        group_copy.drop(columns=["origin_timestamp"], inplace=True)
+
+        train_part = group_copy.iloc[:-horizon]
+        test_part = group_copy.iloc[-horizon:]
+        df_test_part_exog = df_full_part_exog.iloc[-horizon:]
+
+        df_full = pd.concat([df_full, group_copy])
+        df_train = pd.concat([df_train, train_part])
+        df_test = pd.concat([df_test, test_part])
+        df_full_exog = pd.concat([df_full_exog, df_full_part_exog])
+        df_test_exog = pd.concat([df_test_exog, df_test_part_exog])
+
+    TSDataset.to_dataset(df_full).to_csv(
+        dataset_dir / f"m3_{dataset_freq.lower()}_full.csv.gz", index=True, compression="gzip"
+    )
+    TSDataset.to_dataset(df_train).to_csv(
+        dataset_dir / f"m3_{dataset_freq.lower()}_train.csv.gz", index=True, compression="gzip"
+    )
+    TSDataset.to_dataset(df_test).to_csv(
+        dataset_dir / f"m3_{dataset_freq.lower()}_test.csv.gz", index=True, compression="gzip"
+    )
+    TSDataset.to_dataset(df_full_exog).to_csv(
```

Comment:
In theory in can lead to collisions in names. E.g. if we'll have part with name `full_exog` or smth like that. In practice we should just pay attention to such possibility. I don't think we should actively do smth about it.

Reply:
paths look like ../m3_montly/m3_monthly_full_exog
each dataset has its own directory. I don't think we can get collision

Reply:
I think we should try to create a subdirectory `exog` there we will store exog dataframes.